### PR TITLE
fix: add full error handling after os.Stat

### PR DESCRIPTION
**Pull request by dolanor**

### CODE REVIEW

**Comment by gerhard** [comment]
Do you think that we should add a test which shows when this is useful?

The change looks good to me, but I don't understand which situations this "fixes" 🙂

**Comment by dolanor** [comment]
`os.Stat` can fail for different reason. Out of the top of my mind, it could be a permission issue, a network latency (building on some nfs style FS). If we don't catch those, it will try to continue to add the path to the context, which would fail later in the chain with harder to debug issue, I think.

@helderco, you seem to have work most on this file, WDYT?

**Comment by dolanor** [comment]
I guess I could add a test, but since it deals with IO, it might need some refacto. Doesn't have to be big, but I can't say for sure.

**Comment by dolanor** [comment]
Reading a little more, maybe we could use [fstest](https://pkg.go.dev/testing/fstest) if we used `fs.Stat` instead of `os.Stat`. That could be an improvement that would allow injecting mocked FS for testing.
I don't know the performance trade-off. We could benchmark.

**Review by TomChv — COMMENTED** [review]
LGTM

An idea of improvement for later :
Is it not better if we create variables for those error?

For instance
```go
var (
    ErrCannotStat = fmt.Errorf("...")
    ErrNotDir = fmt.Errorf("...")
)
```

**Comment by helderco** [comment]
> Is it not better if we create variables for those error?

I don't think there's a need for that if you're only using them in one place.

**Comment by helderco** [comment]
We don't have plans on exposing a Go API. Only CUE is public 🙂

File: plan/task/clientfilesystemread.go [hunk 1]
```diff
@@ -35,6 +35,8 @@ func (t clientFilesystemReadTask) PreRun(_ context.Context, pctx *plancontext.Co
 		return fmt.Errorf("path %q is not a directory", path)
 	case pi.IsDir() && !isFS:
 		return fmt.Errorf("path %q cannot be a directory", path)
+	case err != nil:
+		return fmt.Errorf("path %q cannot be stat'd", path)
```
Comment (aluzzardi) [inline plan/task/clientfilesystemread.go:39]: Raise original error so user has a chance to fix it (e.g. permission error or whatnot):

```
return fmt.Errorf("path %q cannot be stat'd: %w", path, err)
```

Comment (dolanor) [reply plan/task/clientfilesystemread.go:39]: Right, I forgot to wrap the error. The other cases didn't have any error to wrap…

[the remark this answers is not in the document]
Comment (dolanor) [reply plan/task/clientfilesystemread.go:39]: You're totally right ( https://github.com/dagger/dagger/pull/2018#discussion_r842010500 )
I fixed it.

**Review by aluzzardi — APPROVED** [review]

**Pull request merged**